File: internal/core/push.go
package core

import (
	"context"
	"fmt"
	"path/filepath"

	"git-sfs/internal/cache"
	"git-sfs/internal/errs"
	"git-sfs/internal/hash"
	"git-sfs/internal/lock"
)

type PushOptions struct {
	// SkipMissing uploads the files that are cached locally instead of failing
	// on the first one that is not. Off by default: a push that silently omits
	// files looks identical to a complete one, and a user who then trusts the
	// remote as a backup can lose the omitted data.
	SkipMissing bool
}

// Push uploads each referenced cache file to the remote in a single rclone call.
// Existing remote files are never overwritten.
//
// When a path is given, only symlinks below that path are uploaded. This mirrors
// Pull and is what makes a partially-pulled dataset pushable: subtrees the user
// deliberately never fetched are dangling symlinks, and scanning them would
// abort the push on a cache file that was never supposed to be local.
func (a App) Push(ctx context.Context, name, path string) error {
	return a.PushWithOptions(ctx, name, path, PushOptions{})
}

// PushWithOptions uploads referenced cache files below path to the remote.
func (a App) PushWithOptions(ctx context.Context, name, path string, opts PushOptions) (err error) {
	a.debugf("push: start remote=%s path=%s skip_missing=%t", name, path, opts.SkipMissing)
	defer a.debugDone("push", &err)
	s, err := a.open()
	if err != nil {
		return err
	}
	repo, c, cfg := s.repo, s.cache, s.cfg
	l, err := lock.Acquire(ctx, c.LocksDir(), "push")
	if err != nil {
		return err
	}
	defer l.Release()
	r, err := a.selectRemote(s, name)
	if err != nil {
		return err
	}
	if err := a.preflight(ctx, cfg, r); err != nil {
		return err
	}
	links, err := collectGitSFSSymlinks(repo, path)
	if err != nil {
		return err
	}
	present, missing := partitionByCachePresence(ctx, c, uniqueHashesFromTracked(links))
	if len(missing) > 0 && !opts.SkipMissing {
		return missingCachedFileError(repo, links, missing[0])
	}
	a.reportSkipped(repo, links, missing)
	relPaths := make([]string, 0, len(present))
	for _, h := range present {
		relPaths = append(relPaths, hash.Algorithm+"/"+h.Prefix()+"/"+h.String())
	}
	if len(relPaths) > 0 {
		a.say(fmt.Sprintf("push: uploading %d file(s) to remote", len(relPaths)))
	}
	return r.CopyToRemote(ctx, filepath.Join(c.Root, "files"), relPaths)
}

// partitionByCachePresence splits hashes into those backed by a valid local
// cache file and those that are absent or corrupt, preserving input order so
// the reported results are deterministic.
func partitionByCachePresence(ctx context.Context, c cache.Cache, hashes []hash.Hash) (present, missing []hash.Hash) {
	for _, h := range hashes {
		if c.HasValid(ctx, h) {
			present = append(present, h)
			continue
		}
		missing = append(missing, h)
	}
	return present, missing
}

// maxSkippedListed caps the per-path listing. A partially-pulled dataset can
// have thousands of dangling symlinks, and printing all of them would bury the
// upload result; git-sfs status prints the full list on demand.
const maxSkippedListed = 10

// reportSkipped names the paths left out of the upload. Skipping is opt-in, but
// it still makes the remote an incomplete copy, so the omission is written to
// stderr where it survives being piped and cannot be mistaken for success.
//
// Counts follow git-sfs status: unique cached files, then the symlinks that
// reference them. The two differ whenever paths share content, and reporting
// only the object count would understate how much of the tree is unbacked.
func (a App) reportSkipped(repo string, links []trackedLink, missing []hash.Hash) {
	if len(missing) == 0 {
		return
	}
	affected := linksReferencing(links, missing)
	fmt.Fprintf(a.Stderr,
		"git-sfs: warning: push skipped %d file(s) referenced by %d symlink(s); the remote is not a complete copy\n",
		len(missing), len(affected))
	for i, l := range affected {
		if i == maxSkippedListed {
			fmt.Fprintf(a.Stderr, "  ... and %d more (run: git-sfs status to list all)\n", len(affected)-i)
			break
		}
		fmt.Fprintf(a.Stderr, "  %s (%s)\n", rel(repo, l.Path), l.Hash.Short())
	}
	fmt.Fprintf(a.Stderr, "  run: git-sfs pull <path> to restore them\n")
}

// missingCachedFileError names a working-tree path that references h, so the
// user sees a file they recognize and the command that restores it rather than
// a bare hash. links is sorted by path, so the reported path is deterministic.
func missingCachedFileError(repo string, links []trackedLink, h hash.Hash) error {
	for _, l := range links {
		if l.Hash != h {
			continue
		}
		p := rel(repo, l.Path)
		return fmt.Errorf("%w: %s (%s): run: git-sfs pull %s", errs.ErrMissingCachedFile, p, h.Short(), p)
	}
	return fmt.Errorf("%w: %s", errs.ErrMissingCachedFile, h.Short())
}

// linksReferencing returns every link whose hash is in hashes, preserving the
// sorted-by-path order of links. Reporting has to expand objects back into paths
// because a single cached file can back any number of symlinks, and the user
// needs to know which of their files are affected, not which objects.
func linksReferencing(links []trackedLink, hashes []hash.Hash) []trackedLink {
	want := make(map[hash.Hash]bool, len(hashes))
	for _, h := range hashes {
		want[h] = true
	}
	var out []trackedLink
	for _, l := range links {
		if want[l.Hash] {
			out = append(out, l)
		}
	}
	return out
}

File: crates/git-sfs-core/src/plan/push.rs
//! Pure planning for `push`.
//!
//! Whether an object is actually reachable on the remote is not this
//! module's concern: push always attempts every locally-present object, and
//! `ports::remote`'s `--ignore-existing` is what makes that cheap when the
//! remote already has it. Planning only decides *which local objects*
//! qualify for that attempt.

use std::collections::BTreeSet;

use camino::Utf8PathBuf;
use thiserror::Error;

use crate::domain::hash::Sha256;

use super::TrackedLink;

/// What `push` should do, once local cache presence is known.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct PushPlan {
    /// Unique objects to upload, sorted by hash for deterministic output.
    pub upload: Vec<Sha256>,
    /// Objects left out because they are not cached locally — populated only
    /// under `--skip-missing`; otherwise `plan_push` fails outright instead
    /// (see [`PlanPushError::MissingCachedFile`]).
    pub skipped: Vec<SkippedObject>,
}

/// One object `push --skip-missing` left out, and every tracked path that
/// references it. Reporting only the object total understates how much of the
/// tree is unbacked, since one object can back many paths.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct SkippedObject {
    /// The absent object.
    pub hash: Sha256,
    /// Every tracked path referencing `hash`, in the same order [`plan_push`]
    /// received `links` in (sorted by path, per [`super::TrackedLink`]'s
    /// contract).
    pub paths: Vec<Utf8PathBuf>,
}

/// Why [`plan_push`] refused to produce a plan.
#[derive(Debug, Error, Clone, PartialEq, Eq)]
pub enum PlanPushError {
    /// A referenced cache object is not present locally and `--skip-missing`
    /// was not given. Names a working-tree path, not a bare hash, so the user
    /// sees a file they recognize and the exact command that restores it.
    #[error("cache file missing for {path} ({hash}): run: git-sfs pull {path}")]
    MissingCachedFile {
        /// The path named in the error.
        path: Utf8PathBuf,
        /// The missing object it references.
        hash: Sha256,
    },
}

/// Decides what `push` uploads.
///
/// `links` is the tracked-symlink subset of a [`crate::ports::Repo::scan`]
/// within the requested scope, sorted by path; `present` is every hash
/// `links` references that [`crate::ports::Store::verified`] already
/// confirmed present. A hash in `links` but not in `present` is missing
/// locally.
///
/// # Errors
///
/// Returns [`PlanPushError::MissingCachedFile`] if something is missing and
/// `skip_missing` is `false`.
pub fn plan_push(
    links: &[TrackedLink],
    present: &BTreeSet<Sha256>,
    skip_missing: bool,
) -> Result<PushPlan, PlanPushError> {
    let mut upload = BTreeSet::new();
    let mut missing = BTreeSet::new();
    for link in links {
        if present.contains(&link.hash) {
            upload.insert(link.hash);
        } else {
            missing.insert(link.hash);
        }
    }

    if !missing.is_empty() && !skip_missing {
        let first_missing = links
            .iter()
            .find(|link| missing.contains(&link.hash))
            .expect("missing is non-empty, so some link must reference one of its hashes");
        return Err(PlanPushError::MissingCachedFile {
            path: first_missing.path.clone(),
            hash: first_missing.hash,
        });
    }

    let skipped = missing
        .into_iter()
        .map(|hash| SkippedObject {
            hash,
            paths: links
                .iter()
                .filter(|link| link.hash == hash)
                .map(|link| link.path.clone())
                .collect(),
        })
        .collect();

    Ok(PushPlan {
        upload: upload.into_iter().collect(),
        skipped,
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    fn hash(byte: u8) -> Sha256 {
        Sha256::from_digest([byte; 32])
    }

    fn link(path: &str, hash: Sha256) -> TrackedLink {
        TrackedLink {
            path: Utf8PathBuf::from(path),
            hash,
        }
    }

    #[test]
    fn uploads_every_present_hash_when_nothing_is_missing() {
        let links = vec![link("a.bin", hash(1)), link("b.bin", hash(2))];
        let present = BTreeSet::from([hash(1), hash(2)]);

        let plan = plan_push(&links, &present, false).unwrap();

        assert_eq!(plan.upload, vec![hash(1), hash(2)]);
        assert!(plan.skipped.is_empty());
    }

    #[test]
    fn fails_naming_the_first_missing_path_in_sorted_order_when_skip_missing_is_false() {
        // Links must already be path-sorted (Repo::scan's contract); "a.bin"
        // sorts before "b.bin", so its hash is the one named even though
        // both are missing.
        let links = vec![link("a.bin", hash(1)), link("b.bin", hash(2))];
        let present = BTreeSet::new();

        let err = plan_push(&links, &present, false).unwrap_err();
        assert_eq!(
            err,
            PlanPushError::MissingCachedFile {
                path: Utf8PathBuf::from("a.bin"),
                hash: hash(1),
            }
        );
    }

    #[test]
    fn skip_missing_uploads_present_objects_and_reports_every_path_referencing_a_skipped_one() {
        let links = vec![
            link("present.bin", hash(1)),
            link("dup-a.bin", hash(2)),
            link("dup-b.bin", hash(2)), // shares hash(2) with dup-a.bin
        ];
        let present = BTreeSet::from([hash(1)]);

        let plan = plan_push(&links, &present, true).unwrap();

        assert_eq!(plan.upload, vec![hash(1)]);
        assert_eq!(
            plan.skipped,
            vec![SkippedObject {
                hash: hash(2),
                paths: vec![
                    Utf8PathBuf::from("dup-a.bin"),
                    Utf8PathBuf::from("dup-b.bin")
                ],
            }]
        );
    }

    #[test]
    fn upload_is_deduplicated_by_hash_not_by_path() {
        // Two different paths, same content -- a legitimate dedup case, not
        // an error.
        let links = vec![link("a.bin", hash(1)), link("b.bin", hash(1))];
        let present = BTreeSet::from([hash(1)]);

        let plan = plan_push(&links, &present, false).unwrap();
        assert_eq!(plan.upload, vec![hash(1)]);
    }

    #[test]
    fn an_empty_link_list_produces_an_empty_plan() {
        let plan = plan_push(&[], &BTreeSet::new(), false).unwrap();
        assert!(plan.upload.is_empty());
        assert!(plan.skipped.is_empty());
    }
}
